In [3]:
import io
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
df = pd.read_csv("/content/drive/MyDrive/lotto_df2.csv")

In [10]:
df= df.set_index(['cust', 'clac_mcls_nm'])
df.head()

buy_am  buy_ct  ma_fem_dv  ages  zon_hlv  weat  \
cust       clac_mcls_nm                                                    
M000034966 스포츠잡화           100.0       1          1    45        0     2   
           남성스포츠화        59000.0       1          1    45        0     2   
           남성스포츠화        19000.0       1          1    45        0     2   
           여성스포츠화        62100.0       1          1    45        0     2   
           과채음료           1200.0       1          1    45        0     2   

                         de_hr_1  total_items  holi  weekday  rebuy  \
cust       clac_mcls_nm                                               
M000034966 스포츠잡화               1           63     0        1      0   
           남성스포츠화              1           63     0        1      0   
           남성스포츠화              1           63     0        1      0   
           여성스포츠화              1           63     0        1      0   
           과채음료                1           63     0        1      0   

                         rebuy_diff  rebuy_cnt  rebuy_ration  dt_diff  
cust       clac_mcls_nm                                                
M000034966 스포츠잡화              365.0         34     53.968254    365.0  
           남성스포츠화             365.0         34     53.968254      0.0  
           남성스포츠화             365.0         34     53.968254      0.0  
           여성스포츠화             365.0         34     53.968254      0.0  
           과채음료               365.0         34     53.968254      0.0

- 모델링

In [11]:
X = df.drop(['rebuy'], axis=1)
y = df['rebuy']

In [12]:
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

In [14]:
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.4,random_state=2)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(2615547, 14)
(2615547,)
(1743699, 14)
(1743699,)


In [15]:
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold, cross_val_score, GridSearchCV
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.linear_model import ElasticNet, Lasso, LinearRegression
from sklearn.preprocessing import RobustScaler
import statsmodels.api as sm
from catboost import CatBoostRegressor, Pool
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import HistGradientBoostingRegressor
from ngboost import NGBRegressor

- NMAE
  + NMAE (Normalized Mean Absolute Error)는 척도가 다른 데이터 세트의 MAE에 관한 비교를 용이하게 하기 위해서 정규화를 적용한 것 입니다.
  + 정규화의 수단으로는 측정된 데이터의 평균을 사용합니다.
  + NMAE는 정규화된 MAE(평균 절대 오차) 입니다. 

In [16]:
from sklearn.metrics import make_scorer

def NMAE(true, pred):
    mae = np.mean(np.abs(true - pred))
    score = mae / np.mean(np.abs(true))
    return score

nmae_score = make_scorer(NMAE, greater_is_better=False)

### 모델 탐색


In [ ]:
lr=LinearRegression()
rg=Ridge()
xgb=XGBRegressor(random_state = 42, learning_rate=0.05, n_estimators=20)
gbr=GradientBoostingRegressor(random_state = 42, max_depth = 4, learning_rate=0.01, loss='huber', n_estimators=50)
cat=CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.01, verbose = 0)
lgb=LGBMRegressor(random_state = 42, learning_rate=0.01, n_estimators=50)
ada=AdaBoostRegressor(random_state = 42, learning_rate=0.03, n_estimators=25)
hgb=HistGradientBoostingRegressor(random_state=42, scoring=nmae_score, learning_rate=0.01)

In [ ]:
lr.fit(X_train,y_train)
rg.fit(X_train,y_train)
xgb.fit(X_train,y_train)
gbr.fit(X_train,y_train)
cat.fit(X_train,y_train)
lgb.fit(X_train,y_train)
ngb.fit(X_train,y_train)
ada.fit(X_train,y_train)

In [ ]:
y_lr=lr.predict(X_test)
y_rg=rg.predict(X_test)
y_xgb=xgbpredict(X_test)
y_gbr=gbr.predict(X_test)
y_cat=cat.predict(X_test)
y_lgb=lgb.predict(X_test)
y_ngb=ngb.predict(X_test)
y_ada=ada.predict(X_test)

In [ ]:
score_lr=NMAE(y_test, y_lr)
score_rg=NMAE(y_test, y_rg)
score_xgb=NMAE(y_test, y_xgb)
score_gbr=NMAE(y_test, y_gbr)
score_cat=NMAE(y_test, y_cat)
score_lgb=NMAE(y_test, y_lgb)
score_ngb=NMAE(y_test, y_ngb)
score_ada=NMAE(y_test, y_ada)

In [ ]:
print("LinearRegression: " , score_lr)
print("Ridge: ", score_rg)
print("XGBRegressor: ", score_xgb)
print("GradientBoostingRegressor: ", score_gbr)
print("CatBoostRegressor: ", score_cat)
print("LGBMRegresso: ", score_lgb)
print("AdaBoostRegressor: ", score_ada)
print("HistGradientBoostingRegressor: ", score_hgb)

LinearRegression:  0.9908761860939492
Ridge:  0.9908761960219227
XGBRegressor:  1.0316192485002538
GradientBoostingRegressor:  0.9800517449268821
CatBoostRegressor:  0.81047300041396
LGBMRegresso:  1.050470415863544
AdaBoostRegressor:  0.9838832141340442
HistGradientBoostingRegressor:  1.009584872274517


- CatBoost를 최종 모델로 선택

### 하이퍼파라미터 튜닝

In [20]:
cat_reg=CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.01, verbose = 0)
cat_reg.fit(X_train,y_train)
y_train_pred = cat_reg.predict(X_train)
y_test_pred = cat_reg.predict(X_test)
score_cat=NMAE(y_test, y_test_pred)
print(score_cat)

0.7309049724249184


In [17]:
cat_reg=CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.03, verbose = 0)
cat_reg.fit(X_train,y_train)
y_train_pred = cat_reg.predict(X_train)
y_test_pred = cat_reg.predict(X_test)
score_cat=NMAE(y_test, y_test_pred)
print(score_cat)

0.5901606745120612


In [21]:
cat_reg=CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.01, verbose = 0)
cat_reg.fit(X_train,y_train)
y_train_pred = cat_reg.predict(X_train)
y_test_pred = cat_reg.predict(X_test)
score_cat=NMAE(y_test, y_test_pred)
print(score_cat)

0.5085993982487769


In [25]:
cat_reg=CatBoostRegressor(depth = 16, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.03, verbose = 0)
cat_reg.fit(X_train,y_train)
y_train_pred = cat_reg.predict(X_train)
y_test_pred = cat_reg.predict(X_test)
score_cat=NMAE(y_test, y_test_pred)
print(score_cat)

0.4817188797877027


|하이퍼파라미터|NAME|
|---|---|
|depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.01, verbose = 0|0.730|
|depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.03, verbose = 0|0.590|
|depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.01, verbose = 0)|0.508|
|depth = 16, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.03, verbose = 0)| 0.417|


- 최종 하이퍼파라미터
- depth = 16, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.03, verbose = 0

### 정확도

In [26]:
y_test_preds = [1 if x >= 0.5 else 0 for x in y_test_pred]
print(y_test_preds[:10])

y_train_preds = [1 if x >= 0.5 else 0 for x in y_train_pred]
print(y_train_preds[:10])

[1, 0, 0, 0, 0, 0, 0, 1, 1, 0]
[1, 1, 1, 1, 1, 1, 0, 0, 0, 0]


In [19]:
# CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 100, learning_rate = 0.03, verbose = 0)
print('훈련 MSE: %.3f, 테스트 MSE: %.3f' % (
      mean_squared_error(y_train, y_train_pred),
      mean_squared_error(y_test, y_test_pred)))
print('훈련 R^2: %.3f, 테스트 R^2: %.3f' % (
      r2_score(y_train, y_train_pred),
      r2_score(y_test, y_test_pred)))

print('훈련 정확도: %.3f, 테스트 정확도: %.3f' % (
     accuracy_score(y_train, y_train_preds),
      accuracy_score(y_test, y_test_preds)))

훈련 MSE: 0.217, 테스트 MSE: 0.217
훈련 R^2: 0.120, 테스트 R^2: 0.119
훈련 정확도: 0.742, 테스트 정확도: 0.741


In [24]:
# CatBoostRegressor(depth = 4, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.01, verbose = 0)
print('훈련 MSE: %.3f, 테스트 MSE: %.3f' % (
      mean_squared_error(y_train, y_train_pred),
      mean_squared_error(y_test, y_test_pred)))
print('훈련 R^2: %.3f, 테스트 R^2: %.3f' % (
      r2_score(y_train, y_train_pred),
      r2_score(y_test, y_test_pred)))

print('훈련 정확도: %.3f, 테스트 정확도: %.3f' % (
     accuracy_score(y_train, y_train_preds),
      accuracy_score(y_test, y_test_preds)))

훈련 MSE: 0.175, 테스트 MSE: 0.175
훈련 R^2: 0.290, 테스트 R^2: 0.290
훈련 정확도: 0.783, 테스트 정확도: 0.782


In [27]:
# CatBoostRegressor(depth = 16, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.03, verbose = 0)
print('훈련 MSE: %.3f, 테스트 MSE: %.3f' % (
      mean_squared_error(y_train, y_train_pred),
      mean_squared_error(y_test, y_test_pred)))
print('훈련 R^2: %.3f, 테스트 R^2: %.3f' % (
      r2_score(y_train, y_train_pred),
      r2_score(y_test, y_test_pred)))

print('훈련 정확도: %.3f, 테스트 정확도: %.3f' % (
     accuracy_score(y_train, y_train_preds),
      accuracy_score(y_test, y_test_preds)))

훈련 MSE: 0.190, 테스트 MSE: 0.192
훈련 R^2: 0.228, 테스트 R^2: 0.222
훈련 정확도: 0.790, 테스트 정확도: 0.789


- 최종 정확도: 0.789

#### 앙상블

In [28]:
kf = KFold(n_splits = 10, random_state = 42, shuffle = True)

In [29]:
target = df['rebuy']

In [ ]:
# cat
cat_pred = np.zeros(target.shape[0])
cat_val = []
for n, (tr_idx, val_idx) in enumerate(kf.split(X, y)) :
    tr_x, tr_y = X.iloc[tr_idx], y.iloc[tr_idx]
    val_x, val_y = X.iloc[val_idx], np.expm1(y.iloc[val_idx])
    
    tr_data = Pool(data = tr_x, label = tr_y)
    val_data = Pool(data = val_x, label = val_y)
    
    cat=CatBoostRegressor(depth = 16, random_state = 42, loss_function = 'MAE', n_estimators = 1000, learning_rate = 0.03, verbose = 0)
    cat.fit(tr_data, eval_set = val_data, early_stopping_rounds = 750, verbose = 1000)
    
    val_pred = np.expm1(cat.predict(val_x))
    val_nmae = NMAE(val_y, val_pred)
    cat_val.append(val_nmae)
    print(f'{n + 1} FOLD NMAE = {val_nmae}')
    
    target_data = Pool(data = target, label = None)
    fold_pred = cat.predict(target) / 10
    cat_pred += fold_pred
print(f'10FOLD Mean of NMAE = {np.mean(cat_val)}')